### PR TITLE
[SPARK-18112] [SQL] Support reading data from Hive 2.1 metastore

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?
This PR is to support reading data from Hive 2.1 metastore. Need to update shim class because of the Hive API changes caused by the following three Hive JIRAs:
- [HIVE-12730 MetadataUpdater: provide a mechanism to edit the basic statistics of a table (or a partition)](https://issues.apache.org/jira/browse/HIVE-12730)
- [Hive-13341 Stats state is not captured correctly: differentiate load table and create table](https://issues.apache.org/jira/browse/HIVE-13341)
- [HIVE-13622 WriteSet tracking optimizations](https://issues.apache.org/jira/browse/HIVE-13622)

There are three new fields added to Hive APIs.  
- `boolean hasFollowingStatsTask`. We always set it to `false`. This is to keep the existing behavior unchanged (starting from 0.13), no matter which Hive metastore client version users choose. If we set it to `true`, the basic table statistics is not collected by Hive. For example,

```SQL
	CREATE TABLE tbl AS SELECT 1 AS a
```
When setting `hasFollowingStatsTask ` to `false`, the table properties is like 
```
	Properties: [numFiles=1, transient_lastDdlTime=1489513927, totalSize=2]
```
When setting `hasFollowingStatsTask ` to `true`, the table properties is like 
```
	Properties: [transient_lastDdlTime=1489513563]
```

- `AcidUtils.Operation operation`. Obviously, we do not support ACID. Thus, we set it to `AcidUtils.Operation.NOT_ACID`.
- `EnvironmentContext environmentContext`. So far, this is always set to `null`. This was introduced for supporting DDL `alter table s update statistics set ('numRows'='NaN')`. Using this DDL, users can specify the statistics. So far, our Spark SQL does not need it, because we use different table properties to store our generated statistics values. However, when Spark SQL issues ALTER TABLE DDL statements, Hive metastore always automatically invalidate the Hive-generated statistics. 

In the follow-up PR, we can fix it by explicitly adding a property to `environmentContext`.
```JAVA
putToProperties(StatsSetupConst.STATS_GENERATED, StatsSetupConst.USER)
```
Another alternative is to set `DO_NOT_UPDATE_STATS`to `TRUE`. See the Hive JIRA: https://issues.apache.org/jira/browse/HIVE-15653. We will not address it in this PR.

### How was this patch tested?
Added test cases to VersionsSuite.scala

### CODE REVIEW

**Comment by gatorsmile** [comment]
... The end-to-end test cases in `VersionSuite.scala` does not use the expected hive client versions. Need to fix that at first.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala [hunk 1]
```diff
@@ -95,6 +95,7 @@ private[hive] object IsolatedClientLoader extends Logging {
     case "1.1" | "1.1.0" => hive.v1_1
     case "1.2" | "1.2.0" | "1.2.1" => hive.v1_2
     case "2.0" | "2.0.0" | "2.0.1" => hive.v2_0
+    case "2.1" | "2.1.0" => hive.v2_1
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: Hi, @gatorsmile .
It's great. Could we add "2.1.1" in this PR, too?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/package.scala [hunk 2]
```diff
@@ -67,7 +67,11 @@ package object client {
       exclusions = Seq("org.apache.curator:*",
         "org.pentaho:pentaho-aggdesigner-algorithm"))
 
-    val allSupportedHiveVersions = Set(v12, v13, v14, v1_0, v1_1, v1_2, v2_0)
+    case object v2_1 extends HiveVersion("2.1.0",
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/package.scala]: 2.1.1 here, too?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: I plan to do it later. Thus, I mark it as WIP. The PR is not ready for review

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 3]
```diff
@@ -94,6 +94,10 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     try {
       body
     } catch {
+      case i: InvocationTargetException if isClientException(i.getTargetException) =>
+        val e = i.getTargetException
+        throw new AnalysisException(
+          e.getClass.getCanonicalName + ": " + e.getMessage, cause = Some(e))
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: This fix needs to be backported to the previous releases. Will submit a separate one for this.

**Comment by gatorsmile** [comment]
To show the impact of `boolean hasFollowingStatsTask`, we need to deliver the fix in `VersionSuite.scala` at first. The PR has been submitted: https://github.com/apache/spark/pull/17260

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: The PR is submitted https://github.com/apache/spark/pull/17265

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by gatorsmile** [comment]
cc @cloud-fan @vanzin @sameeragarwal @yhuai @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
Thank you for doing this, @gatorsmile !  It looks good to me.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 4]
```diff
@@ -120,10 +121,12 @@ class VersionsSuite extends SparkFunSuite with Logging {
       System.gc() // Hack to avoid SEGV on some JVM versions.
       val hadoopConf = new Configuration()
       hadoopConf.set("test", "success")
-      // Hive changed the default of datanucleus.schema.autoCreateAll from true to false since 2.0
-      // For details, see the JIRA HIVE-6113
-      if (version == "2.0") {
+      // Hive changed the default of datanucleus.schema.autoCreateAll from true to false and
+      // hive.metastore.schema.verification from false to true since 2.0
+      // For details, see the JIRA HIVE-6113 and HIVE-12463
+      if (version == "2.0" || version == "2.1") {
```
Comment (vanzin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: Is there an issue with setting these values regardless of the version? Adding more and more versions to this whitelist feels sub-optimal.

**Review by vanzin — COMMENTED** [review]
LGTM.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala (hunk 4 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: Either way is fine to me, but `datanucleus.schema.autoCreateAll` was introduced in Hive 2.0 with HIVE-6113 to replace `datanucleus.autoCreateSchema`. That is the major reason why I keep it.

Thanks for the review!

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 5]
```diff
@@ -898,3 +924,107 @@ private[client] class Shim_v2_0 extends Shim_v1_2 {
...
+      loadPath: Path,
+      tableName: String,
+      partSpec: JMap[String, String],
+      replace: Boolean,
+      inheritTableSpecs: Boolean,
+      isSkewedStoreAsSubdir: Boolean,
+      isSrcLocal: Boolean): Unit = {
+    loadPartitionMethod.invoke(hive, loadPath, tableName, partSpec, replace: JBoolean,
+      inheritTableSpecs: JBoolean, isSkewedStoreAsSubdir: JBoolean,
+      isSrcLocal: JBoolean, JBoolean.FALSE, JBoolean.FALSE)
```
Comment (sameeragarwal) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: nit: From a readability perspective, would it make sense to explicitly define `isAcid` and `hasFollowingStatsTask` and set them to `JBoolean.FALSE` (along with some comments for the latter)? We can then use them here and in all the methods below.

**Review by sameeragarwal — APPROVED** [review]
LGTM, thanks!

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: Agree. Let me fix all of them. Thanks!

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**